forgottenserver/src/zones.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+#include <unordered_set>
+#include "tile.h"
+#include "creature.h"
+
+class Zones
+{
+	public:
+		Zones(Tile* tile)
+		{
+			tiles.insert(tile);
+		}
+
+		const std::vector<Position> getPositions() const
+		{
+			std::vector<Position> positions;
+			for (auto& itTile : tiles)
+			{
+				positions.push_back(itTile->getPosition());
+			}
+
+			return positions;
+		}
+
+		void addTile(Tile* tile) { tiles.insert(tile); }
+
+		size_t getCreatureCount(CreatureType_t dataType)	const { return getCount(dataType); }
+		size_t getCreatureCount()							const { return getCount(CREATURETYPE_ANY); }
+		size_t getPlayerCount()								const { return getCount(CREATURETYPE_PLAYER); }
+		size_t getNpcCount()								const { return getCount(CREATURETYPE_NPC); }
+		size_t getMonsterCount()							const { return getCount(CREATURETYPE_MONSTER); }
+		size_t getTilesCount()								const { return tiles.size(); }
+
+		const CreatureVector getCreatures(CreatureType_t dataType)		const { return getCreaturesByType(dataType); }
+		const CreatureVector getCreatures()								const { return getCreaturesByType(CREATURETYPE_ANY); }
+		const CreatureVector getPlayers()								const { return getCreaturesByType(CREATURETYPE_PLAYER); }
+		const CreatureVector getNpcs()									const { return getCreaturesByType(CREATURETYPE_NPC); }
+		const CreatureVector getMonsters()								const { return getCreaturesByType(CREATURETYPE_MONSTER); }
+		const std::unordered_set<Tile*>& getTiles()						const { return tiles; }
+
+	private:
+		inline bool checkCreature(CreatureType_t dataType, const Creature* creature) const
+		{
+			return (dataType == CREATURETYPE_PLAYER && creature->getPlayer()) ||
+				(dataType == CREATURETYPE_NPC && creature->getNpc()) ||
+				(dataType == CREATURETYPE_MONSTER && creature->getMonster());
+		}
+
+		size_t getCount(CreatureType_t dataType) const
```

Comment:
This can be slow with bigger zones. Store creatures when they enter zone and remove when they leave?